src/main/scala/com/courseWork/auth/AuthService.scala

```diff
@@ -0,0 +1,103 @@
+package com.courseWork.auth
+
+import cats.syntax.apply._
+import cats.syntax.applicative._
+import cats.syntax.flatMap._
+import cats.{Applicative, FlatMap, Functor}
+import cats.data.{Validated, ValidatedNec}
+
+import java.util.UUID
+
+trait AuthService[F[_]] {
+  def register(nickname: String, email: String, password: String): F[User]
+  def login(data: LoginData): F[Token]
+}
+
+class AuthServiceImpl[F[_]: FlatMap: Applicative: Functor](userRepository: UserRepository[F]) extends AuthService[F] {
+  type UserValidationErrorsOr[A] = ValidatedNec[UserValidationError, A]
+
+  override def register(nickname: String, email: String, password: String): F[User] = {
+    validateAndGet(nickname, email, password) match {
+      case Validated.Valid(user)         => userRepository.add(user)
+      case Validated.Invalid(errorChain) => ??? // ae.raiseError() // todo: Make MyOwnErrors
+    }
+  }
+
+  override def login(data: LoginData): F[Token] = Token("tokentokentokentokentokentokentokentoken").pure[F]
+
+  private def validateNickname(nickname: String): UserValidationErrorsOr[String] =
+    Validated.condNec[UserValidationError, String](nickname.nonEmpty, nickname, NicknameIsEmpty) *>
+      Validated.condNec(nickname.length > 32, nickname, NicknameTooLong)
+
+  private def validateEmail(email: String): F[UserValidationErrorsOr[String]] = {
+    def matchesRegex(email: String): UserValidationErrorsOr[String] =
+      Validated.condNec[UserValidationError, String](
+        email.matches(
+          "(([a-z0-9][a-z0-9-\\._]*[a-z0-9])|[a-z0-9])@(([a-z0-9][a-z0-9-\\._]*[a-z0-9])|[a-z0-9])\\.([a-z0-9]+)"
+        ),
+        email,
+        IncorrectEmail
+      )
+
+    def userDoesntExist(email: String): F[UserValidationErrorsOr[String]] =
+      userRepository.find(email).flatMap { uOpt =>
+        uOpt
+          .fold[UserValidationErrorsOr[String]](
+            Validated.validNec(email)
+          )(_ => Validated.invalidNec(UserAlreadyExist))
+          .pure[F]
+      }
+
+    matchesRegex(email)
+      .pure[F]
+      .flatMap(matchesValidated =>
+        userDoesntExist(email)
+          .flatMap(userExistsValidate =>
+            (userExistsValidate *> matchesValidated).pure[F]
```

Comment:
I want to use `.map` here, and dont wanna use pure. why `.map` isn't available to me?

Reply:
The whole structure doesn't look clear, I guess the problem with imports and conflicts between abstractions (mentioned in other comments)